src/iop/toneequal.c

```diff
@@ -0,0 +1,3280 @@
+
+  // Simple utils blendings
+  p.details = DT_TONEEQ_GUIDED;
+
+  p.blending = 12.5f;
+  p.feathering = 5.0f;
+  p.iterations = 3;
+  dt_gui_presets_add_generic(_("mask blending : landscapes"), self->op, self->version(), &p, sizeof(p), 1);
+
+  p.blending = 17.67f;
+  p.feathering = 3.53f;
+  p.iterations = 2;
+  dt_gui_presets_add_generic(_("mask blending : all purposes"), self->op, self->version(), &p, sizeof(p), 1);
+
+  p.blending = 25.0f;
+  p.feathering = 2.5f;
+  p.iterations = 1;
+  dt_gui_presets_add_generic(_("mask blending : isolated subjects"), self->op, self->version(), &p, sizeof(p), 1);
+
+  // Shadows/highlights presets
+
+  p.blending = 12.5f;
+  p.iterations = 2;
+  p.quantization = 0.25f;
+  p.feathering = 5.0f;
+  p.method = DT_TONEEQ_NORM_2;
+  p.smoothing = sqrtf(2.0f);
+
+  p.noise = 0.1f;
+  p.ultra_deep_blacks = 0.1f;
+  p.deep_blacks = 0.35f;
+  p.blacks = 0.5f;
+  p.shadows = 0.75f;
+  p.midtones = 0.5f;
+  p.highlights = 0.3f;
+  p.whites = -0.3f;
+  p.speculars = - 0.6f;
+
+  dt_gui_presets_add_generic(_("shadows/highlights : soft"), self->op, self->version(), &p, sizeof(p), 1);
+
+  p.noise = 0.1f;
+  p.ultra_deep_blacks = 0.3f;
+  p.deep_blacks = 0.65f;
+  p.blacks = 0.85f;
+  p.shadows = 0.90f;
+  p.midtones = 0.7f;
+  p.highlights = 0.2f;
+  p.whites = -0.6f;
+  p.speculars = - 0.9f;
+
+  dt_gui_presets_add_generic(_("shadows/highlights : medium"), self->op, self->version(), &p, sizeof(p), 1);
+
+}
+
+
+/**
+ * Helper functions
+ **/
+
+
+static void hash_set_get(uint64_t *hash_in, uint64_t *hash_out, dt_pthread_mutex_t *lock)
+{
+  // Set or get a hash in a struct the thread-safe way
+  dt_pthread_mutex_lock(lock);
+  *hash_out = *hash_in;
+  dt_pthread_mutex_unlock(lock);
+}
+
+
+static void invalidate_luminance_cache(dt_iop_module_t *self)
+{
+  // Invalidate the private luminance cache and histogram when
+  // the luminance mask extraction parameters have changed
+  dt_iop_toneequalizer_gui_data_t *g = (dt_iop_toneequalizer_gui_data_t *)self->gui_data;
+
+  dt_pthread_mutex_lock(&g->lock);
+  g->max_histogram = 1;
+  //g->luminance_valid = 0;
+  g->histogram_valid = 0;
+  g->thumb_preview_hash = 0;
+  g->ui_preview_hash = 0;
+  dt_pthread_mutex_unlock(&g->lock);
+}
+
+
+static int sanity_check(dt_iop_module_t *self)
+{
+  // If tone equalizer is put after flip/orientation module,
+  // the pixel buffer will be in landscape orientation even for pictures displayed in portrait orientation
+  // so the interactive editing will fail. Disable the module and issue a warning then.
+
+  double position_self = self->iop_order;
+  double position_min = dt_ioppr_get_iop_order(self->dev->iop_order_list, "flip");
+
+  if(position_self < position_min && self->enabled)
+  {
+    dt_control_log(_("tone equalizer needs to be after distorsion modules in the pipeline – disabled"));
+    fprintf(stdout, _("tone equalizer needs to be after distorsion modules in the pipeline – disabled"));
```

Comment:
missing \n at end of message in fprintf()

Reply:
done